File: utils/src/main/java/com/github2136/util/CrashHandler.java
package com.github2136.util;

import android.app.Application;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.os.Build;
import android.os.Looper;
import android.widget.Toast;

import java.io.File;
import java.io.PrintWriter;
import java.io.StringWriter;
import java.util.Map;
import java.util.TreeMap;

/**
 * 日志保存
 */
public class CrashHandler implements Thread.UncaughtExceptionHandler {
    private Application application;
    private static CrashHandler mInstance = new CrashHandler();
    // 系统默认的UncaughtException处理类
    private Thread.UncaughtExceptionHandler mDefaultHandler;
    private StringBuffer sb;
    private TreeMap<String, String> map;
    private CrashHandlerCallback mCallback;

    private CrashHandler() {
    }

    /**
     * 单例模式，保证只有一个CustomCrashHandler实例存在
     */
    public static CrashHandler getInstance() {
        return mInstance;
    }

    public void setCustomCrashHanler(Application application, CrashHandlerCallback mCallback) {
        this.application = application;
        // 获取系统默认的UncaughtException处理器
        mDefaultHandler = Thread.getDefaultUncaughtExceptionHandler();
        // 设置该CrashHandler为程序的默认处理器
        Thread.setDefaultUncaughtExceptionHandler(this);
        this.mCallback = mCallback;
    }

    /**
     * 异常发生时，系统回调的函数，我们在这里处理一些操作
     */
    @Override
    public void uncaughtException(Thread thread, Throwable ex) {
        if (!handleException(ex) && mDefaultHandler != null) {
            // 如果用户没有处理则让系统默认的异常处理器来处理
            mDefaultHandler.uncaughtException(thread, ex);
        } else {
            if (BuildConfig.DEBUG && mDefaultHandler != null) {
                mDefaultHandler.uncaughtException(thread, ex);
            }
            try {
                Thread.sleep(4000);
            } catch (InterruptedException e) {
            }
            if (mCallback != null) {
                mCallback.finishAll();
            }
        }
    }

    /**
     * 为我们的应用程序设置自定义Crash处理
     */

    private boolean handleException(final Throwable ex) {
        if (ex == null) {
            return false;
        }
        // 使用Toast来显示异常信息
        new Thread() {
            @Override
            public void run() {
                Looper.prepare();
                Toast.makeText(application, "很抱歉,程序出现异常,即将退出.", Toast.LENGTH_SHORT)
                        .show();
                Looper.loop();
            }
        }.start();
        saveException(application, ex);
        if (mCallback != null) {
            mCallback.submitLog(map, sb.toString());
        }
        return true;
    }

    /**
     * 保存log
     */
    private void saveException(Application application, Throwable ex) {
        StringBuffer sb = getLog(application, ex);
        String filename = FileUtil.createFileName("log", ".log");
        File logFile = new File(FileUtil.getExternalStoragePrivateLogPath(application), filename);
        FileUtil.saveFile(logFile.getPath(), sb.toString());
    }

    private StringBuffer getLog(Application application, Throwable ex) {
        sb = new StringBuffer();
        for (Map.Entry<String, String> entry : getDeviceInfo(application).entrySet()) {
            String key = entry.getKey();
            String value = entry.getValue();
            sb.append(key).append(" = ").append(value).append("\n");
        }
        sb.append(obtainExceptionInfo(ex));
        return sb;
    }

    /**
     * 获取一些简单的信息,软件版本，手机版本，型号等信息存放在HashMap中
     */
    private TreeMap<String, String> getDeviceInfo(Application application) {
        map = new TreeMap<>();
        PackageManager mPackageManager = application.getPackageManager();
        PackageInfo mPackageInfo = null;
        try {
            mPackageInfo = mPackageManager.getPackageInfo(application.getPackageName(), PackageManager.GET_ACTIVITIES);
        } catch (PackageManager.NameNotFoundException e) {
            e.printStackTrace();
        }
        if (mPackageInfo != null) {
            map.put("versionName", mPackageInfo.versionName);
            map.put("versionCode", "" + mPackageInfo.versionCode);
        }
        map.put("SDK_INT", "" + Build.VERSION.SDK_INT);
        map.put("RELEASE", "" + Build.VERSION.RELEASE);
        map.put("CODENAME", "" + Build.VERSION.CODENAME);
        map.put("MODEL", "" + Build.MODEL);
        map.put("PRODUCT", "" + Build.PRODUCT);
        map.put("MANUFACTURER", "" + Build.MANUFACTURER);
        map.put("FINGERPRINT", "" + Build.FINGERPRINT);
        return map;
    }

    /**
     * 获取系统未捕捉的错误信息
     */
    private String obtainExceptionInfo(Throwable throwable) {
        StringWriter mStringWriter = new StringWriter();
        PrintWriter mPrintWriter = new PrintWriter(mStringWriter);
        throwable.printStackTrace(mPrintWriter);
        mPrintWriter.close();
        return mStringWriter.toString();
    }

    public interface CrashHandlerCallback {
        void finishAll();

        void submitLog(Map<String, String> deviceInfo, String exception);
    }
}

File: utils/src/main/java/com/github2136/util/CrashHandler.kt
package com.github2136.util

import android.app.Application
import android.content.pm.PackageInfo
import android.content.pm.PackageManager
import android.os.Build
import android.os.Looper
import android.os.Process
import android.widget.Toast
import java.io.File
import java.io.PrintWriter
import java.io.StringWriter
import kotlin.concurrent.thread

/**
 * Created by yb on 2018/9/4.
 */
class CrashHandler private constructor(val application: Application, val debug: Boolean) : Thread.UncaughtExceptionHandler {
    // 系统默认的UncaughtException处理类
    private var mDefaultHandler: Thread.UncaughtExceptionHandler = Thread.getDefaultUncaughtExceptionHandler()
    private var info: HashMap<String, String> //设备及应用信息
    private var callback: CrashHandlerCallback? = null

    init {
        // 设置该CrashHandler为程序的默认处理器
        Thread.setDefaultUncaughtExceptionHandler(this)
        //设施信息
        info = HashMap()
        try {
            val mPackageManager = application.packageManager
            val mPackageInfo: PackageInfo = mPackageManager.getPackageInfo(application.packageName, PackageManager.GET_ACTIVITIES)
            //APP显示的版本名
            info[APP_VERSION_NAME] = mPackageInfo.versionName
            //APP的版本编号
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
                info[APP_VERSION_CODE] = "${mPackageInfo.longVersionCode}"
            } else {
                info[APP_VERSION_CODE] = "${mPackageInfo.versionCode}"
            }
        } catch (e: PackageManager.NameNotFoundException) {
            e.printStackTrace()
        }

        //Android版本Int
        info[SYS_RELEASE_CODE] = "${Build.VERSION.SDK_INT}"
        //Android显示的版本号
        info[SYS_RELEASE_NAME] = Build.VERSION.RELEASE
        //ABI
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            info[ABIS] = Build.SUPPORTED_ABIS?.contentToString() ?: ""
        } else {
            info[ABIS] = mutableListOf<String>().apply {
                if (Build.CPU_ABI.isNotEmpty()) {
                    add(Build.CPU_ABI)
                }
                if (Build.CPU_ABI2.isNotEmpty()) {
                    add(Build.CPU_ABI2)
                }
            }.joinToString(prefix = "[", postfix = "]") {
                it
            }
        }
        //主板名称
        info[BOARD] = Build.BOARD
        //手机品牌
        info[BRAND] = Build.BRAND
        //工业品设计外观名称，名称没什么规律
        info[DEVICE] = Build.DEVICE
        //显示给用户的名称
        info[DISPLAY] = Build.DISPLAY
        //设备指纹，由厂商、型号、等内容组成的字符，一般同一型号同一系统手机这个值相同
        info[FINGERPRINT] = Build.FINGERPRINT
        //硬件名称，一般为CPU名称
        info[HARDWARE] = Build.HARDWARE
        //变更列表编号，可能是品牌和设备名
        info[ID] = Build.ID
        //硬件制造商
        info[MANUFACTURER] = Build.MANUFACTURER
        //产品具体型号
        info[MODEL] = Build.MODEL
    }

    override fun uncaughtException(t: Thread?, e: Throwable?) {
        val exception = e?.let { obtainExceptionInfo(it) } ?: ""
        callback?.addParam(info)
        saveException(exception)
        callback?.submitLog(info, exception)
        if (debug) {
            mDefaultHandler.uncaughtException(t, e)
        } else {
            // 使用Toast来显示异常信息
            thread {
                Looper.prepare()
                Toast.makeText(application, "很抱歉,程序出现异常,即将退出.", Toast.LENGTH_SHORT).show()
                Looper.loop()
            }
            try {
                Thread.sleep(4000)
            } catch (e: InterruptedException) {
            }
            callback?.finishAll()
            Process.killProcess(Process.myPid())
        }
    }

    /**
     * 保存log
     */
    private fun saveException(ex: String) {
        val sb = StringBuffer()
        info.map { entry -> sb.append(entry.key).append(" = ").append(entry.value).append("\n") }
        val filename = FileUtil.createFileName("log", ".txt")
        //存储至外部私有目录
        File(FileUtil.getExternalStoragePrivateLogPath(application), filename).writeText(sb.toString() + ex)
    }

    /**
     * 获取系统未捕捉的错误信息
     */
    private fun obtainExceptionInfo(throwable: Throwable): String {
        val mStringWriter = StringWriter()
        val mPrintWriter = PrintWriter(mStringWriter)
        throwable.printStackTrace(mPrintWriter)
        mPrintWriter.close()
        return mStringWriter.toString()
    }

    fun setCallback(lis: CrashHandlerCallback) {
        callback = lis
    }

    interface CrashHandlerCallback {
        /**
         * 添加KV参数
         */
        fun addParam(info: HashMap<String, String>)
        /**
         * 结束所有界面
         */
        fun finishAll()

        /**
         * 提交保存异常
         */
        fun submitLog(info: Map<String, String>, exception: String)
    }

    companion object {
        @Volatile
        private var instance: CrashHandler? = null

        //APP的版本编号
        const val APP_VERSION_CODE = "APP_VERSION_CODE"
        //APP显示的版本名
        const val APP_VERSION_NAME = "APP_VERSION_NAME"
        //Android版本Int
        const val SYS_RELEASE_CODE = "SYS_RELEASE_CODE"
        //Android显示的版本号
        const val SYS_RELEASE_NAME = "SYS_RELEASE_NAME"
        //ABI
        const val ABIS = "ABIS"
        //主板名称
        const val BOARD = "BOARD"
        //手机品牌
        const val BRAND = "BRAND"
        //工业品设计外观名称，名称没什么规律
        const val DEVICE = "DEVICE"
        //显示给用户的名称
        const val DISPLAY = "DISPLAY"
        //设备指纹，由厂商、型号、等内容组成的字符，一般同一型号同一系统手机这个值相同
        const val FINGERPRINT = "FINGERPRINT"
        //硬件名称，一般为CPU名称
        const val HARDWARE = "HARDWARE"
        //变更列表编号，可能是品牌和设备名
        const val ID = "ID"
        //硬件制造商
        const val MANUFACTURER = "MANUFACTURER"
        //产品具体型号
        const val MODEL = "MODEL"

        fun getInstance(application: Application, debug: Boolean): CrashHandler {
            if (instance == null) {
                synchronized(CrashHandler::class) {
                    if (instance == null) {
                        instance = CrashHandler(application, debug)
                    }
                }
            }
            return instance!!
        }
    }
}